ow_regolith/src/RegolithSpawner.cpp

```diff
@@ -0,0 +1,295 @@
+    return false;
+  }
+  Vector3 gravity(
+    msg.response.gravity.x, msg.response.gravity.y, msg.response.gravity.z
+  );
+  // set psuedo force to the model's weight multiplied by the weight factor
+  m_psuedo_force_mag 
+    = model_mass * gravity.length() * PSUEDO_FORCE_WEIGHT_FACTOR;
+
+  // subscribe callbacks to ROS topics
+  m_modify_terrain_visual = m_node_handle->subscribe(
+    TOPIC_MODIFY_TERRAIN_VISUAL, 1, &RegolithSpawner::terrainVisualModCb, this);
+  m_dig_linear_result = m_node_handle->subscribe(
+    TOPIC_DIG_LINEAR_RESULT, 1, &RegolithSpawner::armDigLinearResultCb, this);
+  m_dig_circular_result = m_node_handle->subscribe(
+    TOPIC_DIG_CIRCULAR_RESULT, 1, &RegolithSpawner::armDigCircularResultCb, this);
+  m_deliver_result = m_node_handle->subscribe(
+    TOPIC_DELIVER_RESULT, 1, &RegolithSpawner::armDeliverResultCb, this);
+
+  return true;
+}
+
+bool RegolithSpawner::spawnRegolithInScoop(Vector3 pushback_direction)
+{
+  ROS_INFO("Spawning regolith");
+
+  // spawn model
+  static auto spawn_count = 0;
+  stringstream model_name;
+  model_name << "regolith_" << spawn_count++;
+
+  SpawnModel spawn_msg;
+  
+  spawn_msg.request.model_name                  = model_name.str();
+  spawn_msg.request.model_xml                   = m_model_sdf;
+  spawn_msg.request.robot_namespace             = "/regolith";
+  spawn_msg.request.reference_frame             = "lander::l_scoop_tip";
+
+  spawn_msg.request.initial_pose.position.x     = 0.0;
+  spawn_msg.request.initial_pose.position.y     = 0.0;
+  spawn_msg.request.initial_pose.position.z     = -0.05;
+  
+  spawn_msg.request.initial_pose.orientation.x  = 0.0;
+  spawn_msg.request.initial_pose.orientation.y  = 0.0;
+  spawn_msg.request.initial_pose.orientation.z  = 0.0;
+  spawn_msg.request.initial_pose.orientation.w  = 0.0;
+  
+  if (!callRosService(m_gz_spawn_model, spawn_msg))
+    return false;
+
+  // apply a force to keep model in the scoop
+  stringstream body_name;
+  body_name << model_name.str() << "::" << m_model_linkname;
+
+  ApplyBodyWrench wrench_msg;
+  
+  wrench_msg.request.body_name         = body_name.str();
+  wrench_msg.request.reference_frame   = "world";
+  
+  wrench_msg.request.reference_point.x = 0.0;
+  wrench_msg.request.reference_point.y = 0.0;
+  wrench_msg.request.reference_point.z = 0.0;
+  
+  tf::vector3TFToMsg(
+    m_psuedo_force_mag * pushback_direction, 
+    wrench_msg.request.wrench.force
+  );
+  
+  wrench_msg.request.wrench.torque.x   = 0.0;
+  wrench_msg.request.wrench.torque.y   = 0.0;
+  wrench_msg.request.wrench.torque.z   = 0.0;
+  
+  // negative duration means the force is applied until we clear it
+  wrench_msg.request.duration          = ros::Duration(-1);
+
+  m_active_models.push_back({model_name.str(), body_name.str()});
+
+  return callRosService(m_gz_apply_wrench, wrench_msg);
+}
+
+
+void RegolithSpawner::clearAllPsuedoForces()
+{
+  for (auto &regolith : m_active_models) {
+    BodyRequest msg;
```

Comment:
I think this would be slightly more efficient (unless the compiler is smart enough):
```suggestion
{
  BodyRequest msg;
  for (auto &regolith : m_active_models) {
```

Note this is okay since BodayRequest has only one field that you overwrite on each iteration